### PR TITLE
Replaying games

### PR DESCRIPTION
**Pull request by dalabayev**
Additional information is added to save in JSON.

Additional methods: replayGame and createSavedStatistics are added.

Lines are replaced by circle when drawing eye movements.

Some bug are fixed.

### CODE REVIEW

File: gazeplay-commons/src/main/java/net/gazeplay/commons/gaze/devicemanager/AbstractGazeDeviceManager.java [hunk 1]
```diff
@@ -170,6 +170,39 @@ synchronized void onGazeUpdate(Point2D gazePositionOnScreen) {
 
         }
     }
+    @Override
+    public void onSavedMovementsUpdate(Point2D gazePositionOnScreen) {
```
Comment (kuhlkrein) [inline gazeplay-commons/src/main/java/net/gazeplay/commons/gaze/devicemanager/AbstractGazeDeviceManager.java:174]: I think it's same as onGazeUpdate ? (probably for testing), please don't forget to delete when tests will be finished

File: gazeplay-commons/src/main/java/net/gazeplay/commons/utils/games/ImageLibrary.java [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package net.gazeplay.commons.utils.games;
 
 import javafx.scene.image.Image;
+import net.gazeplay.commons.random.ReplayablePseudoRandom;
```
Comment (kuhlkrein) [inline gazeplay-commons/src/main/java/net/gazeplay/commons/utils/games/ImageLibrary.java:4]: Is it useful here ?

File: gazeplay-commons/src/main/java/net/gazeplay/commons/utils/stats/Stats.java [hunk 3]
```diff
@@ -33,9 +37,7 @@
 import java.awt.image.BufferedImageOp;
 import java.awt.image.ConvolveOp;
 import java.awt.image.Kernel;
-import java.io.File;
-import java.io.IOException;
-import java.io.PrintWriter;
+import java.io.*;
```
Comment (kuhlkrein) [inline gazeplay-commons/src/main/java/net/gazeplay/commons/utils/stats/Stats.java:40]: try to use the less number of imports possible

File: gazeplay-commons/src/main/java/net/gazeplay/commons/utils/stats/Stats.java [hunk 4]
```diff
@@ -433,8 +463,17 @@ public void start() {
...
             recordGazeMovements = e -> {
+                final long timeToFixation = System.currentTimeMillis() - startTime;
+                final long timeInterval = (timeToFixation - previousTime);
                 final int getX = (int) e.getX();
                 final int getY = (int) e.getY();
+                JsonObject savedDataObj2 = new JsonObject();
+                savedDataObj2.addProperty("X", getX);
+                savedDataObj2.addProperty("Y", getY);
+                savedDataObj2.addProperty("time", timeInterval);
+                //saveCoordinates("{\"X\":" + getX + ", \"Y\":" + getY + ", \"time\":" + timeInterval + "}");
```
Comment (kuhlkrein) [inline gazeplay-commons/src/main/java/net/gazeplay/commons/utils/stats/Stats.java:474]: remove comments after tests

File: gazeplay-core/src/main/java/net/gazeplay/GazePlay.java [hunk 5]
```diff
@@ -26,6 +26,11 @@
 import org.springframework.context.annotation.Lazy;
 import org.springframework.stereotype.Component;
 
+import javax.swing.*;
```
Comment (kuhlkrein) [inline gazeplay-core/src/main/java/net/gazeplay/GazePlay.java:29]: check imports and comments after tests

**Review by kuhlkrein — COMMENTED** [review]
few comments

File: gazeplay-commons/src/main/java/net/gazeplay/commons/gaze/devicemanager/AbstractGazeDeviceManager.java (hunk 1 above)
Comment (dalabayev) [reply gazeplay-commons/src/main/java/net/gazeplay/commons/gaze/devicemanager/AbstractGazeDeviceManager.java:175]: yes, you're right. it's the same. I didn't want to change onGazeUpdate method and I left onSavedMovementsUpdate method and added to it the call for onGazeUpdate.

File: gazeplay-commons/src/main/java/net/gazeplay/commons/utils/games/ImageLibrary.java (hunk 2 above)
Comment (dalabayev) [reply gazeplay-commons/src/main/java/net/gazeplay/commons/utils/games/ImageLibrary.java]: Done.

File: gazeplay-commons/src/main/java/net/gazeplay/commons/utils/stats/Stats.java (hunk 3 above)
Comment (dalabayev) [reply gazeplay-commons/src/main/java/net/gazeplay/commons/utils/stats/Stats.java]: ok. thank you.

File: gazeplay-commons/src/main/java/net/gazeplay/commons/utils/stats/Stats.java (hunk 4 above)
Comment (dalabayev) [reply gazeplay-commons/src/main/java/net/gazeplay/commons/utils/stats/Stats.java]: Done.

File: gazeplay-core/src/main/java/net/gazeplay/GazePlay.java (hunk 5 above)
Comment (dalabayev) [reply gazeplay-core/src/main/java/net/gazeplay/GazePlay.java]: Done.

**Comment by kuhlkrein** [comment]
Please have a look at the continuous integration error:

Dm: Reliance on default encoding (DM_DEFAULT_ENCODING)
Found a call to a method which will perform a byte to String (or String to byte) conversion, and will assume that the default platform encoding is suitable. This will cause the application behaviour to vary between platforms. Use an alternative API and specify a charset name or Charset object explicitly.

```
  <BugInstance type="DM_DEFAULT_ENCODING" priority="1" rank="19" abbrev="Dm" category="I18N">
    <Class classname="net.gazeplay.commons.utils.stats.Stats">
      <SourceLine classname="net.gazeplay.commons.utils.stats.Stats" start="59" end="899" sourcefile="Stats.java" sourcepath="net/gazeplay/commons/utils/stats/Stats.java"/>
    </Class>
    <Method classname="net.gazeplay.commons.utils.stats.Stats" name="saveStats" signature="()Lnet/gazeplay/commons/utils/stats/SavedStatsInfo;" isStatic="false">
      <SourceLine classname="net.gazeplay.commons.utils.stats.Stats" start="580" end="648" startBytecode="0" endBytecode="1494" sourcefile="Stats.java" sourcepath="net/gazeplay/commons/utils/stats/Stats.java"/>
    </Method>
    <Method classname="java.io.FileWriter" name="&lt;init&gt;" signature="(Ljava/io/File;)V" isStatic="false" role="METHOD_CALLED">
      <SourceLine classname="java.io.FileWriter" start="96" end="97" startBytecode="0" endBytecode="68" sourcefile="FileWriter.java" sourcepath="java/io/FileWriter.java"/>
    </Method>
    <SourceLine classname="net.gazeplay.commons.utils.stats.Stats" start="607" end="607" startBytecode="255" endBytecode="255" sourcefile="Stats.java" sourcepath="net/gazeplay/commons/utils/stats/Stats.java"/>
  </BugInstance>
```

**Comment by kuhlkrein** [comment]
@dalabayev Also we have to change the icon for the replay button before merging ^^ Something like this :
image

in the style of the other GazePlay buttons could be good :) I think we have blue green yellow and red colors for now... so maybe purple or orange could be a good choice ?

**Pull request merged**